File: src/App/WebRoot/js/upsert.js
(function () {
    var addTagButton = document.getElementById('Upsert-addTagButton');
    var removeTagButtons = document.getElementsByClassName('Upsert-removeTagButton');
    var ul = document.getElementById('Upsert-tagList');
    var submitButton = document.getElementById('Upsert-submit');
    var select = document.getElementById('Id');

    select.addEventListener('change', function (e, d) {
        e.preventDefault();
        var href = window.location.href.split('?')[0];

        if (e.target.value === '0') {
            window.location.href = href;
            return;
        }

        window.location.href = href + '?id=' + e.target.value;
    })

    submitButton.addEventListener('click', function (e) {
        e.preventDefault();

        var formNames = [
            'Id',
            'ArticleDate',
            'Title',
            'Description',
            'Source',
            'Tags'
        ]

        var isValid = formNames.map(function (fn) {
            var elem = document.getElementsByName(fn);
            if (!elem.length) {
                return false;
            }

            return !!elem[0].value;
        }).reduce(function (acc, cur) {
            return acc && cur;
        }, true)

        if (isValid) {
            document.getElementsByClassName('PostContents')[0].submit();
        } else {
            alert('Form not valid');
        }
    });

    addTagButton.addEventListener('click', function (e) {
        e.preventDefault();
        var div = document.createElement('div');
        div.classList = 'Upsert-tagValue';

        var li = document.createElement('li');
        li.classList = 'Upsert-tagListItem';

        var input = document.createElement('input');
        input.type = 'text';
        input.classList = 'Upsert-tagName';
        input.setAttribute('name', 'Tags');

        var button = document.createElement('input');
        button.type = 'button'
        button.classList = 'Upsert-removeTagButton';
        button.value = '-';
        button.addEventListener('click', removeTag);

        div.appendChild(input);
        div.appendChild(button);

        li.appendChild(div);
        ul.appendChild(li);
    });

    var removeTag = function (e) {
        e.preventDefault();
        var li = this.parentNode.parentNode;
        li.remove();
    }

    for (var i = 0; i < removeTagButtons.length; i++) {
        removeTagButtons[i].addEventListener('click', removeTag);
    }
})();

File: src/App/Views/Upsert.fs
namespace Hashset.Views

open Model

module Upsert =
    open Giraffe.ViewEngine

    let view (upsertDocument: UpsertDocument) availableImages =

        let idDropdown =
            option [ _value "0" ] [ str "New" ] ::
            (upsertDocument.ExistingIds
            |> Seq.sortByDescending (fun (id, _) -> id)
            |> Seq.map (fun (id, title) ->
                let attrs =
                    if title = upsertDocument.Title
                    then _selected :: [ _value (id.ToString()) ]
                    else [ _value (id.ToString()) ]
                option attrs [ str title ]
            )
            |> Seq.toList)

        div [ _class "Upsert-page" ] [
            link [ _rel "stylesheet"; _type "text/css"; _href "/css/upsert.css"; _async ]
            form [ _action "/upsert"; _method "POST"; _class "PostContents Upsert-form"; _enctype "multipart/form-data" ] [
                div [ _class "Upsert-inputRow" ] [
                    label [ _class "Upsert-inputLabel"; _for "Id" ] [ str "Id" ]
                    select [ _class "Upsert-inputValue"; _id "Id"; _name "Id"; ] idDropdown
                ]
                div [ _class "Upsert-inputRow" ] [
                    label [ _class "Upsert-inputLabel"; _for "ArticleDate" ] [ str "ArticleDate" ]
                    input [ _class "Upsert-inputValue"; _type "date"; _id "ArticleDate"; _name "ArticleDate"; _value (upsertDocument.ArticleDate.ToString("yyyy-MM-dd")) ]
                ]
                div [ _class "Upsert-inputRow" ] [
                    label [ _class "Upsert-inputLabel"; _for "Title" ] [ str "Title" ]
                    input [ _class "Upsert-inputValue"; _type "text"; _id "Title"; _name "Title"; _value upsertDocument.Title ]
                ]
                div [ _class "Upsert-inputRow" ] [
                    label [ _class "Upsert-inputLabel"; _for "Description" ] [ str "Description" ]
                    input [ _class "Upsert-inputValue"; _type "text"; _id "Description"; _name "Description"; _value upsertDocument.Description ]
                ]
                div [ _class "Upsert-inputRow" ] [
                    label [ _class "Upsert-inputLabel"; _for "Images" ] [ str "Upload Images" ]
                    input [ _class "Upsert-inputValue Upsert-inputImages"; _type "file"; _id "Images"; _name "Images"; _multiple ]
                ]
                div [] [
                    label [ _class "Upsert-inputSourceLabel"; _for "Source" ] [ str "Source" ]
                    textarea [ _class "Upsert-inputSourceValue"; _rows "10"; _id "Source"; _name "Source" ] [ str upsertDocument.Source ]
                ]
                div [] [
                    div [ _class "Upsert-tagContainer" ] [
                        label [ _class "Upsert-tagLabel" ] [ str "Tags" ]
                        button [ _id "Upsert-addTagButton" ] [ str "+" ]
                    ]
                    ul [ _id "Upsert-tagList" ]
                        (upsertDocument.Tags
                        |> List.map (fun t ->
                            li [ _class "Upsert-tagListItem" ] [
                                div [ _class "Upsert-tagValue" ] [
                                    input [ _type "text"; _name "Tags"; _class "Upsert-tagName"; _value t ]
                                    input [ _type "button"; _class "Upsert-removeTagButton"; _value "-" ]
                                ]
                            ]
                        ))
                ]

                input [ _id "Upsert-submit"; _type "Submit" ]

                script [ _async; _defer; _src "/js/upsert.js"; _type "text/javascript"; ] []
            ]
            aside [ _class "Upsert-availableImagesAside"]
                (availableImages
                 |> List.map (fun ai ->
                     div [ _class "Upsert-availableImageContainer" ] [
                         label [ _class "Upsert-availableImageLabel" ] [ str ai ]
                         img [ _class "Upsert-availableImage"; _src <| sprintf "/images/%s" ai ]
                     ]
                 ))
        ]
